### ISSUE TITLE
Question: Is there a sniff to detect the use of double quotes over single quotes?

### ISSUE THREAD

**Issue by u01jmg3** [squizlabs/PHP_CodeSniffer#509]
- Is there a current sniff to detect double quotes (unless string interpolation is in use) and instead recommend the use of single quotes? Small thing but curious if a sniff existed.
  - https://stackoverflow.com/questions/13620/speed-difference-in-using-inline-strings-vs-concatenation-in-php5#answer-13680

**Comment by aik099** [comment]
Can you please write example of that PHP code. Examples on provided link weren't clear.

**Comment by u01jmg3** [comment]
```php
$s = "these are " . $foo;   //warning
$s = 'these are ' . $foo;   //no warning
$s = "these are $foo";      //no warning
```

**Comment by aik099** [comment]
I saw that example, but since what's in `$x` and `$z` are equivalent, then why not just use single quotes. Also theoretically such string should work faster.

**Comment by u01jmg3** [comment]
Yep, indeed why not just singles quotes and yes it would be faster - anyway PHPCS could be used to enforce this?

**Comment by aik099** [comment]
Then back to original question in issue description: what is string interpolation?

I think sniff properly detects double quote usage, even you your example.

**Comment by u01jmg3** [comment]
Definition: _In computer programming, **string interpolation** or variable substitution is the process of evaluating a string literal containing one or more placeholders, yielding a result in which the placeholders are replaced with their corresponding values._

String interpolation would be the example `$y` whereby double quotes would need to be used and PHPCS should give no warning.This is as well as `$z` giving no warning.

Which available sniff allows me to detect double quote usage as per `$x`?

**Comment by aik099** [comment]
I see, then `Squiz.Strings.DoubleQuoteUsage` is the one you need.

> $x = "these are " . $foo;   //warning

What warning message do you see? If you run `phpcs` with `-s` option, then also sniff code will be displayed.

**Comment by u01jmg3** [comment]
This is just what I'm looking for but it's also disallowing example `$y`. Anyway to tone down how strict PHPCS is and allow it but prevent `$x`?

```php
$x = "these are " . $foo;   //String "these are " does not require double quotes; use single quotes instead
$y = "these are $foo";      //Variable "$foo" not allowed in double quote string; use concatenation instead
$z = 'these are ' . $foo;   //no warning
```

**Comment by gsherwood** [comment]
You can mute that specific error in your ruleset.xml file like this:
```xml
 <rule ref="Squiz.Strings.DoubleQuoteUsage.ContainsVar">
  <severity>0</severity>
 </rule>
```
Just by coincidence, this happens to be the exact example I picked for muting an error message on the annotated ruleset wiki page: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset.xml

:)

Does that do what you are after?

**Comment by u01jmg3** [comment]
Thank you both

@gsherwood: thanks for answering my final question. I thought I was familiar with PHPCS but it seems I need to do some more reading!

**Comment by dbaltas** [comment]
This can be achieved with the following rule
```
<rule ref="Squiz.Strings.DoubleQuoteUsage.NotRequired" />
```
Which does not allow double quotes for plain strings but does allow them if they contain a variable.

**Comment by josephzidell** [comment]
In php 7, interpolation [is actually faster](https://blog.blackfire.io/php-7-performance-improvements-encapsed-strings-optimization.html) than concatenation. Is there a reverse sniff that finds usages of concats-that-could-be-interpolations?

**Comment by gsherwood** [comment]
Maybe you want to comment on #2316

**Comment by josephzidell** [comment]
That one is about the quotes on an existing string. My use-case would be to detect single- or double-quoted strings that are currently concatenated and could use interpolation, which is basically a variable or class property (excludes constants, method calls).

**Comment by gsherwood** [comment]
Yes, but it may be more useful if a sniff that checks for single quote usage instead checks for concatenation. So if a sniff is to be written, I think it's worth putting your comments there for consideration rather than on this closed issue. Issue #2259 is probably also worth watching.

But I didn't answer your original question, which is no, there is no sniff to enforce the reverse that I know of.

**Comment by josephzidell** [comment]
Will do. Thanks
